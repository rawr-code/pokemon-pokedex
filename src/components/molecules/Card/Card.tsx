import { FC } from 'react'
import Image from 'next/image'

// Atoms
import { Icons } from '@atoms'
import { IconsName } from '@/components/atoms/Icons/Icons'

// Utils
import { createTestIDs, getColor, cn } from '@utils'

const ownTestIDs = createTestIDs('Card', [
  'root',
  'pokeball',
  'header',
  'content',
  'types',
  'image',
])

// Types
interface ICardProps {
  name: string
  id: string
  types: IconsName[]
  img?: string
  onClick: () => void

  className?: string
  classes?: {
    root?: string
  }
  testIDs?: { [k in keyof typeof ownTestIDs]?: string | undefined }
}

const Card: FC<ICardProps> = ({
  name,
  id,
  types,
  img,
  onClick,
  className = '',
  classes = {},
  testIDs = {},
}) => {
  const backgroundColor = getColor(types[0])

  return (
    <div
      className={cn(
        'relative h-full w-full cursor-pointer overflow-hidden rounded-xl p-3 pt-2',
        className,
        classes.root,
      )}
      onClick={onClick}
      id={name}
      style={{ backgroundColor }}
      data-testid={testIDs.root || ownTestIDs.root}
    >
      {/* Pokeball */}
      <div
        className="absolute -bottom-4 -right-3 z-10 w-1/2 contrast-100"
        data-testid={testIDs.pokeball || ownTestIDs.pokeball}
      >
        <Icons.pokeball size={96} color="#fff" className="opacity-10" />
      </div>
      {/* Header */}
      <div
        className="flex items-start justify-between text-white"
        data-testid={testIDs.header || ownTestIDs.header}
      >
        <p className="text-md capitalize">{name}</p>
        <span className="text-xs">{id}</span>
      </div>
      {/* Content */}
      <div
        className="relative z-20 mt-2.5 flex justify-between"
        data-testid={testIDs.content || ownTestIDs.content}
      >
        {/* Types */}
        <div
          className="flex flex-col space-y-1"
          data-testid={testIDs.types || ownTestIDs.types}
        >
          {types.map(t => {
            const Icon = Icons[t]
            return (
              <div
                key={t}
                className="flex w-max items-center rounded-full bg-black/20 px-1 py-1"
              >
                <div
                  className="flex h-5 w-5 items-center justify-center rounded-full"
                  style={{ backgroundColor }}
                >
                  <Icon size={12} color="#fff" />
                </div>
                <span className="mx-1 text-sm capitalize text-white">{t}</span>
              </div>
            )
          })}
        </div>
        {/* Image */}
        <div
          className="relative flex h-14 w-14 items-center justify-center"
          data-testid={testIDs.image || ownTestIDs.image}
        >
          {img ? (
            <Image src={img} alt="" className="h-full w-full" fill />
          ) : null}
        </div>
      </div>
    </div>
  )
}

export default Card
export { ownTestIDs as testIDs }
