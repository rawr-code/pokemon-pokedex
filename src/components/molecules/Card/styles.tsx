import React from 'react'
import tw, { css, styled } from 'twin.macro'

export const Wrapper = tw.div`w-1/2 pb-3 pl-3 h-32`

interface StyledContainerProps {
  color: string
  children: React.ReactNode
}

export const StyledContainer = styled.div<StyledContainerProps>`
  ${({ color }) => css`
    background-color: ${color};
  `}
  ${tw`relative w-full cursor-pointer overflow-hidden rounded-xl p-3 pt-2 h-full`}
`

export const Container = ({ color, children }: StyledContainerProps) => (
  <StyledContainer color={color}>{children}</StyledContainer>
)

export const PokeballWrapper = tw.div`absolute -bottom-4 -right-3 z-10 w-1/2 contrast-100`

export const Header = tw.div`flex items-start justify-between text-white`

export const ContentWrapper = tw.div`relative z-20 mt-2.5 flex justify-between`

export const TypesWrapper = tw.div`flex flex-col space-y-1`

export const Bagde = tw.div`flex items-center rounded-full bg-black/20 px-1 py-1`

export const StyledBadgeContainer = styled.div<StyledContainerProps>`
  ${({ color }) => css`
    background-color: ${color};
  `}
  ${tw`flex h-5 w-5 items-center justify-center rounded-full`}
`

export const BadgeContainer = ({ color, children }: StyledContainerProps) => (
  <StyledBadgeContainer color={color}>{children}</StyledBadgeContainer>
)

export const BadgeText = tw.span`mx-1 text-sm capitalize text-white`

export const ImageWrapper = tw.div`relative flex h-14 w-14 items-center justify-center`
