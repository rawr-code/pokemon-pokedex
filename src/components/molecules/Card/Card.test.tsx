import runCommonTests from '@/testing/runCommonTests'

// Component
import Card, { testIDs } from './Card'

describe('Molecule | Card', () => {
  runCommonTests(
    Card,
    { rootTestID: testIDs.root },
    {
      id: '#001',
      name: 'bulbasaur',
      types: ['grass', 'poison'],
      img: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg',
      onClick: () => undefined,
    },
  )
})
