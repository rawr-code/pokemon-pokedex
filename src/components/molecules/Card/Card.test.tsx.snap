// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecule | Card should match snapshot 1`] = `
<div>
  <div
    class="relative h-full w-full cursor-pointer overflow-hidden rounded-xl p-3 pt-2 className classes.root"
    data-testid="Card-root"
    id="bulbasaur"
    style="background-color: rgb(99, 188, 90);"
  >
    <div
      class="absolute -bottom-4 -right-3 z-10 w-1/2 contrast-100"
      data-testid="Card-pokeball"
    >
      <svg
        class="opacity-10"
        fill="none"
        height="96"
        viewBox="0 0 208 208"
        width="96"
      >
        <path
          d="M128.762 104c0 13.676-11.086 24.762-24.762 24.762S79.238 117.676 79.238 104 90.324 79.238 104 79.238 128.762 90.324 128.762 104z"
          fill="#fff"
        />
        <path
          clip-rule="evenodd"
          d="M104 208c52.393 0 95.738-38.743 102.947-89.143h-60.912c-6.118 17.312-22.628 29.714-42.035 29.714-19.407 0-35.916-12.402-42.035-29.714H1.053C8.263 169.257 51.607 208 104 208zM61.965 89.143H1.053C8.263 38.743 51.607 0 104 0s95.738 38.743 102.947 89.143h-60.912C139.917 71.83 123.407 59.429 104 59.429c-19.407 0-35.916 12.402-42.035 29.714zM128.762 104c0 13.676-11.086 24.762-24.762 24.762S79.238 117.676 79.238 104 90.324 79.238 104 79.238 128.762 90.324 128.762 104z"
          fill="#fff"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="flex items-start justify-between text-white"
      data-testid="Card-header"
    >
      <p
        class="text-md capitalize"
      >
        bulbasaur
      </p>
      <span
        class="text-xs"
      >
        #001
      </span>
    </div>
    <div
      class="relative z-20 mt-2.5 flex justify-between"
      data-testid="Card-content"
    >
      <div
        class="flex flex-col space-y-1"
        data-testid="Card-types"
      >
        <div
          class="flex w-max items-center rounded-full bg-black/20 px-1 py-1"
        >
          <div
            class="flex h-5 w-5 items-center justify-center rounded-full"
            style="background-color: rgb(99, 188, 90);"
          >
            <svg
              fill="none"
              height="12"
              viewBox="0 0 130 139"
              width="12"
            >
              <path
                clip-rule="evenodd"
                d="M21.443 121.044a67.122 67.122 0 01-1.466-1.422c-25.505-25.505-25.505-66.857 0-92.362C45.482 1.755 126.751.05 126.751.05s11.093 94.068-14.412 119.573c-22.652 22.652-57.803 25.186-83.251 7.602l20.505-25.163 32.78-7.077-26.758-2.675L72.62 74.994l19.357-4.24-15.307-4.533L91.977 38.61 70.205 63.03l-8.407-11.864 2.96 19.589-15.165 17.095-7.03-21.628v28.763l-21.12 26.06z"
                fill="#fff"
                fill-rule="evenodd"
              />
            </svg>
          </div>
          <span
            class="mx-1 text-sm capitalize text-white"
          >
            grass
          </span>
        </div>
        <div
          class="flex w-max items-center rounded-full bg-black/20 px-1 py-1"
        >
          <div
            class="flex h-5 w-5 items-center justify-center rounded-full"
            style="background-color: rgb(99, 188, 90);"
          >
            <svg
              fill="none"
              height="12"
              viewBox="0 0 144 145"
              width="12"
            >
              <path
                clip-rule="evenodd"
                d="M120.325 111.057C134.866 99.43 144 82.631 144 63.948 144 28.851 111.765.4 72 .4 32.236.4 0 28.85 0 63.948c0 18.008 8.486 34.267 22.122 45.83a33.13 33.13 0 00-1.28 9.203c0 14.038 8.483 25.419 18.948 25.419 6.8 0 12.762-4.805 16.105-12.022C59.237 139.595 65.2 144.4 72 144.4c6.198 0 11.701-3.993 15.158-10.165 3.457 6.172 8.96 10.165 15.158 10.165 10.464 0 18.947-11.381 18.947-25.419 0-2.767-.329-5.43-.938-7.924zm-6.641-45.848c0 17.818-19.087 32.263-42.631 32.263-23.545 0-42.632-14.445-42.632-32.263 0-17.819 19.087-32.263 42.632-32.263 23.544 0 42.631 14.444 42.631 32.263z"
                fill="#fff"
                fill-rule="evenodd"
              />
            </svg>
          </div>
          <span
            class="mx-1 text-sm capitalize text-white"
          >
            poison
          </span>
        </div>
      </div>
      <div
        class="relative flex h-14 w-14 items-center justify-center"
        data-testid="Card-image"
      >
        <img
          alt=""
          class="h-full w-full"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
    </div>
  </div>
</div>
`;
